cpp/src/arrow/extension/fixed_shape_tensor_test.cc

```diff
@@ -0,0 +1,293 @@
+  ASSERT_TRUE(sliced->storage()->Equals(*partial->storage()));
+  ASSERT_EQ(sliced->length(), partial->length());
+}
+
+void CheckSerializationRoundtrip(const std::shared_ptr<ExtensionType>& ext_type) {
+  auto serialized = ext_type->Serialize();
+  ASSERT_OK_AND_ASSIGN(auto deserialized,
+                       ext_type->Deserialize(ext_type->storage_type(), serialized));
+  ASSERT_TRUE(ext_type->Equals(*deserialized));
+}
+
+TEST_F(TestExtensionType, MetadataSerializationRoundtrip) {
+  CheckSerializationRoundtrip(fixed_shape_tensor(value_type_, {}, {}, {}));
+  CheckSerializationRoundtrip(fixed_shape_tensor(value_type_, {0}, {}, {}));
+  CheckSerializationRoundtrip(fixed_shape_tensor(value_type_, {1}, {0}, {"x"}));
+  CheckSerializationRoundtrip(
+      fixed_shape_tensor(value_type_, {256, 256, 3}, {0, 1, 2}, {"H", "W", "C"}));
+  CheckSerializationRoundtrip(
+      fixed_shape_tensor(value_type_, {256, 256, 3}, {2, 0, 1}, {"C", "H", "W"}));
+
+  auto ext_type = fixed_shape_tensor(value_type_, cell_shape_, {0, 1}, dim_names_);
+  CheckSerializationRoundtrip(ext_type_);
+
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid, testing::HasSubstr("Invalid: Expected FixedSizeList storage type"),
+      ext_type->Deserialize(boolean(), serialized_));
+}
+
+TEST_F(TestExtensionType, RoudtripBatch) {
+  auto exact_ext_type = internal::checked_pointer_cast<FixedShapeTensorType>(ext_type_);
+
+  std::vector<std::shared_ptr<Buffer>> buffers = {nullptr, Buffer::Wrap(values_)};
+  auto arr_data = std::make_shared<ArrayData>(value_type_, values_.size(), buffers, 0, 0);
+  auto arr = std::make_shared<Int64Array>(arr_data);
+  EXPECT_OK_AND_ASSIGN(auto fsla_arr, FixedSizeListArray::FromArrays(arr, cell_type_));
+  auto data = fsla_arr->data();
+  data->type = ext_type_;
+  auto ext_arr = exact_ext_type->MakeArray(data);
+
+  ASSERT_OK(UnregisterExtensionType(ext_type_->extension_name()));
+  ASSERT_OK(RegisterExtensionType(ext_type_));
+  auto ext_metadata =
+      key_value_metadata({{"ARROW:extension:name", exact_ext_type->extension_name()},
+                          {"ARROW:extension:metadata", serialized_}});
+  auto ext_field = field("f0", exact_ext_type, true, ext_metadata);
+  auto batch = RecordBatch::Make(schema({ext_field}), ext_arr->length(), {ext_arr});
+  std::shared_ptr<RecordBatch> read_batch;
+  RoundtripBatch(batch, &read_batch);
+  CompareBatch(*batch, *read_batch, /*compare_metadata=*/true);
+}
+
+TEST_F(TestExtensionType, RoudtripBatchFromTensor) {
+  auto exact_ext_type = internal::checked_pointer_cast<FixedShapeTensorType>(ext_type_);
+  ASSERT_OK_AND_ASSIGN(auto tensor, Tensor::Make(value_type_, Buffer::Wrap(values_),
+                                                 shape_, {}, dim_names_));
+  EXPECT_OK_AND_ASSIGN(auto ext_arr, exact_ext_type->MakeArray(tensor));
+  ext_arr->data()->type = exact_ext_type;
+
+  ASSERT_OK(UnregisterExtensionType(ext_type_->extension_name()));
+  ASSERT_OK(RegisterExtensionType(ext_type_));
```

Comment:
What's the reason for unregistering/registering again here?

Reply:
Oh, misunderstood how registration works - I assumed the exact type used had to be registered to serialize/deserialize. Removed.

Reply:
Yes, otherwise the registry couldn't work for users in practice, since you can only register the name once, but there are many possible parametrizations of this type (and which you also don't know in advance while reading data)